File: sdk/typescript/src/format/pq.ts
/**
 * Post-quantum cryptographic operations for CEF.
 *
 * Uses @noble/post-quantum for ML-KEM-768 (FIPS 203) key encapsulation
 * and ML-DSA-65 (FIPS 204) digital signatures.
 *
 * ⚠️  PROTOTYPING ONLY — not for production use.
 * noble-post-quantum is audited but not FIPS-certified. Use the Go SDK
 * with CIRCL for production PQ workloads.
 */

import { ml_kem768 } from '@noble/post-quantum/ml-kem.js';
import { ml_dsa65 } from '@noble/post-quantum/ml-dsa.js';
import { aesKeyWrap, aesKeyUnwrap, sha256, zeroize } from './crypto.js';

const subtle = globalThis.crypto.subtle;

/** Convert Uint8Array to BufferSource for WebCrypto compatibility. */
function buf(data: Uint8Array): BufferSource {
  return data.buffer.slice(data.byteOffset, data.byteOffset + data.byteLength) as ArrayBuffer;
}
import {
  AlgMLKEM768_A256KW, AlgMLDSA65, HeaderKeyID,
  type RecipientInfo, type Recipient,
  type WrapCEKFunc, type UnwrapCEKFunc,
  type SignFunc, type VerifyFunc,
} from './cose.js';

// Re-export algorithm IDs for convenience
export { AlgMLKEM768_A256KW, AlgMLDSA65 };

// Domain separation label — must match Go SDK
const DOMAIN_LABEL = 'CEF-ML-KEM-768-A256KW';

// ---------------------------------------------------------------------------
// ML-KEM-768 key encapsulation
// ---------------------------------------------------------------------------

export interface MLKEMKeyPair {
  publicKey: Uint8Array;  // 1184 bytes
  secretKey: Uint8Array;  // 2400 bytes
}

/**
 * Generate an ML-KEM-768 key pair.
 */
export function mlkemKeygen(): MLKEMKeyPair {
  const { publicKey, secretKey } = ml_kem768.keygen();
  return { publicKey, secretKey };
}

/**
 * Derive a 256-bit KEK from an ML-KEM shared secret using HKDF-SHA256 (RFC 5869).
 *
 * Parameters:
 * - IKM: ML-KEM shared secret (32 bytes)
 * - salt: empty (zero-length)
 * - info: domain label "CEF-ML-KEM-768-A256KW"
 * - length: 32 bytes
 *
 * Must match the Go implementation.
 */
async function deriveMLKEMKEK(sharedSecret: Uint8Array): Promise<Uint8Array> {
  const info = new TextEncoder().encode(DOMAIN_LABEL);

  // Import shared secret as HKDF key material
  const ikm = await subtle.importKey('raw', buf(sharedSecret), 'HKDF', false, ['deriveBits']);

  // HKDF-SHA256: Extract + Expand
  const derived = await subtle.deriveBits(
    { name: 'HKDF', hash: 'SHA-256', salt: new Uint8Array(0), info: buf(info) },
    ikm,
    256, // 32 bytes
  );

  return new Uint8Array(derived);
}

/**
 * Create a WrapCEKFunc that uses ML-KEM-768 encapsulation.
 *
 * The recipient's public key must be provided in RecipientInfo.publicKey
 * (1184 bytes, raw ML-KEM-768 public key).
 */
export function mlkemWrapCEK(recipientPublicKeys: Map<string, Uint8Array>): WrapCEKFunc {
  return async (cek: Uint8Array, recipient: RecipientInfo): Promise<Uint8Array> => {
    const pk = recipientPublicKeys.get(recipient.keyId);
    if (!pk) {
      throw new Error(`mlkem: no public key for recipient ${recipient.keyId}`);
    }

    // Encapsulate — produces ciphertext + shared secret
    const { cipherText, sharedSecret } = ml_kem768.encapsulate(pk);

    // Derive KEK from shared secret with domain separation
    const kek = await deriveMLKEMKEK(sharedSecret);

    // Zeroize shared secret immediately after KEK derivation
    zeroize(sharedSecret);

    // Wrap CEK with derived KEK
    const wrappedCEK = await aesKeyWrap(kek, cek);

    // Zeroize KEK immediately after wrapping
    zeroize(kek);

    // Return ciphertext || wrappedCEK
    // The recipient needs the ML-KEM ciphertext to decapsulate,
    // then the wrapped CEK to unwrap.
    const result = new Uint8Array(cipherText.length + wrappedCEK.length);
    result.set(cipherText);
    result.set(wrappedCEK, cipherText.length);
    return result;
  };
}

/**
 * Create an UnwrapCEKFunc that uses ML-KEM-768 decapsulation.
 */
export function mlkemUnwrapCEK(secretKey: Uint8Array): UnwrapCEKFunc {
  return async (wrappedData: Uint8Array, _recipient: Recipient): Promise<Uint8Array> => {
    // ML-KEM-768 ciphertext is 1088 bytes
    const ctLen = 1088;
    if (wrappedData.length <= ctLen) {
      throw new Error(`mlkem: wrapped data too short (${wrappedData.length} bytes, need >${ctLen})`);
    }

    const cipherText = wrappedData.slice(0, ctLen);
    const wrappedCEK = wrappedData.slice(ctLen);

    // Decapsulate to recover shared secret
    const sharedSecret = ml_kem768.decapsulate(cipherText, secretKey);

    // Derive KEK with same domain separation
    const kek = await deriveMLKEMKEK(sharedSecret);

    // Zeroize shared secret immediately after KEK derivation
    zeroize(sharedSecret);

    // Unwrap CEK
    const cekResult = await aesKeyUnwrap(kek, wrappedCEK);

    // Zeroize KEK immediately after unwrapping
    zeroize(kek);

    return cekResult;
  };
}

// ---------------------------------------------------------------------------
// ML-DSA-65 digital signatures
// ---------------------------------------------------------------------------

export interface MLDSAKeyPair {
  publicKey: Uint8Array;  // 1952 bytes
  secretKey: Uint8Array;  // 4032 bytes
}

/**
 * Generate an ML-DSA-65 key pair.
 */
export function mldsaKeygen(): MLDSAKeyPair {
  const { publicKey, secretKey } = ml_dsa65.keygen();
  return { publicKey, secretKey };
}

/**
 * Create a SignFunc using ML-DSA-65.
 */
export function mldsaSign(secretKey: Uint8Array): SignFunc {
  return async (sigStructure: Uint8Array): Promise<Uint8Array> => {
    return ml_dsa65.sign(sigStructure, secretKey);
  };
}

/**
 * Create a VerifyFunc using ML-DSA-65.
 */
export function mldsaVerify(publicKey: Uint8Array): VerifyFunc {
  return async (sigStructure: Uint8Array, signature: Uint8Array): Promise<void> => {
    const valid = ml_dsa65.verify(signature, sigStructure, publicKey);
    if (!valid) {
      throw new Error('ML-DSA-65: signature verification failed');
    }
  };
}

/**
 * Raw ML-DSA-65 sign (synchronous, for certificate generation).
 * Returns a function that signs a message with the given secret key.
 */
export function mldsaSignRaw(secretKey: Uint8Array): (message: Uint8Array) => Uint8Array {
  return (message: Uint8Array): Uint8Array => {
    return ml_dsa65.sign(message, secretKey);
  };
}

File: sdk/rust/src/format/pq.rs
//! Post-quantum cryptographic primitives.

use ml_kem::MlKem768;
#[allow(deprecated)]
use ml_kem::ExpandedKeyEncoding;
use ml_kem::kem::{Encapsulate, Decapsulate, KeyExport};
use ml_dsa::MlDsa65;
use ml_dsa::signature::{Signer, Verifier, Keypair as _};
use zeroize::Zeroize;

use crate::error::CefError;
use super::crypto::{aes_key_wrap, aes_key_unwrap, hkdf_sha256, zeroize_bytes};

const DOMAIN_LABEL: &[u8] = b"CEF-ML-KEM-768-A256KW";

pub struct MLKEMKeyPair { pub public_key: Vec<u8>, pub secret_key: Vec<u8> }
pub struct MLDSAKeyPair { pub public_key: Vec<u8>, pub secret_key: Vec<u8> }

pub fn mlkem_keygen() -> MLKEMKeyPair {
    use rand::Rng;
    let mut seed = [0u8; 64];
    rand::rng().fill(&mut seed);
    let dk = ml_kem::DecapsulationKey::<MlKem768>::from_seed(ml_kem::Seed::from(seed));
    let ek = dk.encapsulation_key().clone();
    MLKEMKeyPair {
        public_key: ek.to_bytes().to_vec(),
        secret_key: seed.to_vec(),
    }
}

pub fn mldsa_keygen() -> MLDSAKeyPair {
    use getrandom::SysRng;
    use rand_core::UnwrapErr;
    let sk: ml_dsa::SigningKey<MlDsa65> = <MlDsa65 as ml_dsa::KeyGen>::key_gen(&mut UnwrapErr(SysRng));
    let vk = sk.verifying_key().clone();
    let seed = sk.to_seed();
    MLDSAKeyPair {
        public_key: vk.encode().to_vec(),
        secret_key: seed.to_vec(),
    }
}

fn derive_kek(ss: &[u8]) -> Result<Vec<u8>, CefError> {
    hkdf_sha256(ss, &[], DOMAIN_LABEL, 32)
}

pub fn mlkem_wrap(public_key: &[u8], cek: &[u8]) -> Result<Vec<u8>, CefError> {
    if public_key.len() != 1184 {
        return Err(CefError::Crypto(format!("ML-KEM pk: {} bytes, need 1184", public_key.len())));
    }
    let ek_arr: &ml_kem::array::Array<u8, _> = public_key.try_into()
        .map_err(|_| CefError::Crypto("invalid ML-KEM pk".into()))?;
    let ek = ml_kem::EncapsulationKey::<MlKem768>::new(ek_arr)
        .map_err(|e| CefError::Crypto(format!("ML-KEM pk: {e}")))?;

    let (ct, mut ss) = ek.encapsulate_with_rng(&mut rand_core::UnwrapErr(getrandom::SysRng));
    let mut kek = derive_kek(ss.as_ref())?;
    ss.zeroize();
    let wrapped = aes_key_wrap(&kek, cek)?;
    zeroize_bytes(&mut kek);
    let ct_ref: &[u8] = ct.as_ref();
    let mut out = Vec::with_capacity(ct_ref.len() + wrapped.len());
    out.extend_from_slice(ct_ref);
    out.extend_from_slice(&wrapped);
    Ok(out)
}

pub fn mlkem_unwrap(secret_key: &[u8], wrapped_data: &[u8]) -> Result<Vec<u8>, CefError> {
    // Accept both 64-byte seed and 2400-byte expanded key (interop with Go/TS)
    let dk = match secret_key.len() {
        64 => {
            let seed: [u8; 64] = secret_key.try_into().unwrap();
            ml_kem::DecapsulationKey::<MlKem768>::from_seed(ml_kem::Seed::from(seed))
        }
        2400 => {
            #[allow(deprecated)]
            let dk_arr: &ml_kem::ExpandedDecapsulationKey<MlKem768> = secret_key.try_into()
                .map_err(|_| CefError::Crypto("invalid 2400-byte ML-KEM key".into()))?;
            #[allow(deprecated)]
            ml_kem::DecapsulationKey::<MlKem768>::from_expanded_bytes(dk_arr)
                .map_err(|e| CefError::Crypto(format!("ML-KEM expanded key: {e}")))?
        }
        n => return Err(CefError::Crypto(format!("ML-KEM sk: {} bytes, need 64 or 2400", n))),
    };
    let ct_size = 1088usize;
    if wrapped_data.len() <= ct_size {
        return Err(CefError::Crypto(format!("wrapped data too short: {}", wrapped_data.len())));
    }
    let ct: &ml_kem::Ciphertext<MlKem768> = wrapped_data[..ct_size].try_into()
        .map_err(|_| CefError::Crypto("invalid ciphertext".into()))?;
    let wrapped_cek = &wrapped_data[ct_size..];
    let mut ss = dk.decapsulate(ct);
    let mut kek = derive_kek(ss.as_ref())?;
    ss.zeroize();
    let cek = aes_key_unwrap(&kek, wrapped_cek)?;
    zeroize_bytes(&mut kek);
    Ok(cek)
}

pub fn mldsa_sign(secret_key: &[u8], message: &[u8]) -> Result<Vec<u8>, CefError> {
    if secret_key.len() != 32 {
        return Err(CefError::Crypto(format!("ML-DSA seed: {} bytes, need 32", secret_key.len())));
    }
    let seed: [u8; 32] = secret_key.try_into().unwrap();
    let sk = <MlDsa65 as ml_dsa::KeyGen>::from_seed(&ml_dsa::Seed::from(seed));
    let sig = sk.sign(message);
    Ok(sig.encode().to_vec())
}

pub fn mldsa_verify(public_key: &[u8], message: &[u8], signature: &[u8]) -> Result<(), CefError> {
    if public_key.len() != 1952 {
        return Err(CefError::Crypto(format!("ML-DSA pk: {} bytes, need 1952", public_key.len())));
    }
    if signature.len() != 3309 {
        return Err(CefError::Crypto(format!("ML-DSA sig: {} bytes, need 3309", signature.len())));
    }
    let vk_bytes: [u8; 1952] = public_key.try_into().unwrap();
    let vk = ml_dsa::VerifyingKey::<MlDsa65>::decode(
        &ml_dsa::EncodedVerifyingKey::<MlDsa65>::from(vk_bytes)
    );
    let sig_bytes: [u8; 3309] = signature.try_into().unwrap();
    let sig = ml_dsa::Signature::<MlDsa65>::decode(
        &ml_dsa::EncodedSignature::<MlDsa65>::from(sig_bytes)
    ).ok_or_else(|| CefError::Crypto("invalid ML-DSA signature".into()))?;
    vk.verify(message, &sig)
        .map_err(|_| CefError::Crypto("ML-DSA: verification failed".into()))
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test] fn mlkem_round_trip() {
        let kp = mlkem_keygen();
        let cek = crate::format::crypto::random_bytes(32);
        let w = mlkem_wrap(&kp.public_key, &cek).unwrap();
        assert_eq!(mlkem_unwrap(&kp.secret_key, &w).unwrap(), cek);
    }
    #[test] fn mldsa_sign_verify() {
        let kp = mldsa_keygen();
        let sig = mldsa_sign(&kp.secret_key, b"test").unwrap();
        mldsa_verify(&kp.public_key, b"test", &sig).unwrap();
    }
    #[test] fn mldsa_wrong_key() {
        let k1 = mldsa_keygen(); let k2 = mldsa_keygen();
        let sig = mldsa_sign(&k1.secret_key, b"x").unwrap();
        assert!(mldsa_verify(&k2.public_key, b"x", &sig).is_err());
    }
    #[test] fn mlkem_wrong_key() {
        let k1 = mlkem_keygen(); let k2 = mlkem_keygen();
        let cek = crate::format::crypto::random_bytes(32);
        let w = mlkem_wrap(&k1.public_key, &cek).unwrap();
        assert!(mlkem_unwrap(&k2.secret_key, &w).is_err());
    }

    #[test]
    fn mlkem_expanded_key_unwrap() {
        use rand::Rng;
        let mut seed = [0u8; 64];
        rand::rng().fill(&mut seed);
        let dk = ml_kem::DecapsulationKey::<MlKem768>::from_seed(ml_kem::Seed::from(seed));
        let ek = dk.encapsulation_key().clone();
        #[allow(deprecated)]
        let expanded = dk.to_expanded_bytes();

        let pk = ek.to_bytes().to_vec();
        let sk_expanded = expanded.to_vec();
        assert_eq!(sk_expanded.len(), 2400);

        let cek = crate::format::crypto::random_bytes(32);
        let wrapped = mlkem_wrap(&pk, &cek).unwrap();
        let unwrapped = mlkem_unwrap(&sk_expanded, &wrapped).unwrap();
        assert_eq!(unwrapped, cek);
    }
}
